firstRectForCharacterRange returns the original range instead of the actualRange for the first selection rect.

megangardner opened the pull request:
#### c585393d11f122301036b3d1b3cc03a64737afa7
<pre>
firstRectForCharacterRange returns the original range instead of the actualRange for the first selection rect.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274913">https://bugs.webkit.org/show_bug.cgi?id=274913</a>
<a href="https://rdar.apple.com/129013199">rdar://129013199</a>

Reviewed by Wenson Hsieh.

Existing FIXME, pass the actual range of the first
line of the selection instead of just passing back the
range that was passed in.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::firstRectForCharacterRangeAsync):
* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(TEST(WKWebViewMacEditingTests, FirstRectForCharacterRange)):

Canonical link: <a href="https://commits.webkit.org/279576@main">https://commits.webkit.org/279576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/735d58ad88353cf9372da716cd815d6fb09274f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33163 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6311 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57073 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4371 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43560 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2954 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31379 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46525 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24698 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28208 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3836 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2673 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58668 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28961 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4083 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50971 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30156 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46676 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50307 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11729 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31093 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29936 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -285,4 +285,31 @@ - (NSRange)range
+    __unsafe_unretained id<NSTextInputClient_Async> inputClient = (id<NSTextInputClient_Async>)webView.get();
+
+    __block bool done = false;
+    [inputClient firstRectForCharacterRange: { 0, 31 } completionHandler:^(NSRect firstRect, NSRange actualRange) {
```
```suggestion
    [inputClient firstRectForCharacterRange:{ 0, 31 } completionHandler:^(NSRect firstRect, NSRange actualRange) {
```

whsieh commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -285,4 +285,31 @@ - (NSRange)range
+TEST(WKWebViewMacEditingTests, FirstRectForCharacterRange)
+{
+    auto configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"WebProcessPlugInWithInternals" configureJSCForTesting:YES];
+    auto webView = adoptNS([[TestWKWebView<NSTextInputClient, NSTextInputClient_Async> alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration]);
```
```suggestion
    RetainPtr webView = adoptNS([[TestWKWebView<NSTextInputClient, NSTextInputClient_Async> alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration]);
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6909,11 +6909,14 @@ void WebPage::firstRectForCharacterRangeAsync(const EditingRange& editingRange,
+    if (!lineEndBoundary)
+        return completionHandler({ }, { });
 
+    completionHandler(rect, EditingRange::fromRange(*frame, makeSimpleRange(range->start, WTFMove(lineEndBoundary))));
```
It would probably be a good idea to clamp this first line editing range to the given `editingRange`. I think we should also pass the given `editingRange` through, in the case where the selection is collapsed.

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -6907,13 +6907,20 @@ void WebPage::firstRectForCharacterRangeAsync(const EditingRange& editingRange,
+
+    auto rangeForFirstLine = EditingRange::fromRange(*frame, makeSimpleRange(range->start, WTFMove(lineEndBoundary)));
 
+    rangeForFirstLine.length = std::min(rangeForFirstLine.length, editingRange.length);
```
Hm.. I'd expect this to clamp `rangeForFirstLine`, such that its start is at or inside `editingRange`, and the extent is also at or inside `editingRange`. Something kinda like this?

```
rangeForFirstLine.location = std::min(std::max(rangeForFirstLine.location, editingRange.location), editingRange.location + editingRange.length);
rangeForFirstLine.length = std::min(rangeForFirstLine.location + rangeForFirstLine.length, editingRange.location + editingRange.length) - rangeForFirstLine.location;
```

megangardner committed:
firstRectForCharacterRange returns the original range instead of the actualRange for the first selection rect.
https://bugs.webkit.org/show_bug.cgi?id=274913
rdar://129013199

Reviewed by Wenson Hsieh.

Existing FIXME, pass the actual range of the first
line of the selection instead of just passing back the
range that was passed in.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::firstRectForCharacterRangeAsync):
* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(TEST(WKWebViewMacEditingTests, FirstRectForCharacterRange)):

Canonical link: https://commits.webkit.org/279576@main
